### PR TITLE
[API] Organizer can set additional event info

### PR DESCRIPTION
**Pull request by Jon1969Edwards**
## Description
As an Organizer
In order to engage relevant audience
I would like to be able to set additional event info

## PivotalTracker
[Link](https://www.pivotaltracker.com/story/show/163653937)

## Tasks
- Create request specs
- Event attributes to available to endpoint: title, description, location, date & time

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -21,6 +21,6 @@ def create
 
   private
   def event_params
-    params.require(:event).permit(:title)
+    params.require(:event).permit(:title, :description, :location, :date_and_time)
```
Comment (CamronLDNF) [inline app/controllers/events_controller.rb]: The correct naming convention is `:datetime` I believe. Do a google search.

File: app/models/event.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 class Event < ApplicationRecord
-  validates_presence_of :title
+  validates_presence_of :title, :description, :location, :date_and_time
```
Comment (CamronLDNF) [inline app/models/event.rb]: same here with date

File: app/serializers/events/for_group_collection_serializer.rb [hunk 3]
```diff
@@ -1,3 +1,3 @@
 class Events::ForGroupCollectionSerializer < ActiveModel::Serializer
-  attributes :id, :title, :date
+  attributes :id, :title, :date, :description, :location, :date_and_time
```
Comment (CamronLDNF) [inline app/serializers/events/for_group_collection_serializer.rb]: Duplication of the date info

File: app/serializers/events/index_serializer.rb [hunk 4]
```diff
@@ -1,4 +1,4 @@
 class Events::IndexSerializer < ActiveModel::Serializer
-  attributes :id, :title, :date
+  attributes :id, :title, :date, :description, :location, :date_and_time
```
Comment (CamronLDNF) [inline app/serializers/events/index_serializer.rb]: Duplication of the date info

File: db/migrate/20190204200935_add_columns_to_event.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class AddColumnsToEvent < ActiveRecord::Migration[5.2]
+  def change
+    add_column :events, :description, :text
+    add_column :events, :location, :string
+    add_column :events, :date_and_time, :string
```
Comment (CamronLDNF) [inline db/migrate/20190204200935_add_columns_to_event.rb]: Fix date naming  AND  date needs to be `:datetime` data type I suspect. Look it up.

File: db/schema.rb [hunk 6]
```diff
@@ -21,13 +21,30 @@
...
+    t.datetime "run_at"
+    t.datetime "locked_at"
+    t.datetime "failed_at"
+    t.string "locked_by"
+    t.string "queue"
+    t.datetime "created_at"
+    t.datetime "updated_at"
+    t.index ["priority", "run_at"], name: "delayed_jobs_priority"
+  end
+
```
Comment (CamronLDNF) [inline db/schema.rb]: This is a strange addition. Drop and re-create the database.

File: db/schema.rb [hunk 7]
```diff
@@ -21,13 +21,30 @@
...
   create_table "events", force: :cascade do |t|
     t.string "title"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
     t.bigint "group_id"
     t.date "date"
-    t.integer "organizer"
+    t.text "description"
+    t.string "location"
+    t.string "date_and_time"
```
Comment (CamronLDNF) [inline db/schema.rb]: Date naming and data type again

File: spec/factories/events.rb [hunk 8]
```diff
@@ -2,6 +2,9 @@
   factory :event do
     title { "EventName" }
     date { "2019-12-12" }
+    description { "Description for event" }
+    location { "Location for event" }
+    date_and_time { "29 Feb-5 Pm" }
```
Comment (CamronLDNF) [inline spec/factories/events.rb]: Fix date naming and format of example

File: spec/models/event_spec.rb [hunk 9]
```diff
@@ -3,10 +3,16 @@
 RSpec.describe Event, type: :model do
   describe 'DB table' do
     it { is_expected.to have_db_column :title }
+    it { is_expected.to have_db_column :description }
+    it { is_expected.to have_db_column :location }
+    it { is_expected.to have_db_column :date_and_time }
```
Comment (CamronLDNF) [inline spec/models/event_spec.rb]: datetime

File: spec/models/event_spec.rb [hunk 10]
```diff
@@ -3,10 +3,16 @@
...
   end
 
   describe 'Validations' do
     it { is_expected.to validate_presence_of :title }
+    it { is_expected.to validate_presence_of :description }
+    it { is_expected.to validate_presence_of :location }
+    it { is_expected.to validate_presence_of :date_and_time }
```
Comment (CamronLDNF) [inline spec/models/event_spec.rb]: datetime

File: spec/requests/organizer_can_create_event_spec.rb [hunk 11]
```diff
@@ -11,7 +11,12 @@
     let(:headers) { { HTTP_ACCEPT: 'application/json' }.merge!(user_credentials) }
     
     before do
-      post "/groups/#{group.id}/events", params: { event: {title: 'Craft Academy'} }, headers: headers
+      post "/groups/#{group.id}/events", params: { event: 
+                                                  {title: 'Craft Academy',
+                                                  description: 'Graduation Party',
+                                                  location: 'Stockholm',
+                                                  date_and_time: '09 Feb, 1pm',} }, headers: headers
```
Comment (CamronLDNF) [inline spec/requests/organizer_can_create_event_spec.rb]: 1. Fix datetime 
2. Add organizer association
3. Need to test for all the attributes as well

File: spec/requests/organizer_can_create_event_spec.rb [hunk 12]
```diff
@@ -11,7 +11,12 @@
     let(:headers) { { HTTP_ACCEPT: 'application/json' }.merge!(user_credentials) }
```
Comment (CamronLDNF) [inline spec/requests/organizer_can_create_event_spec.rb]: Further up in this file, user ought to be associated to the group as the organizer

File: spec/serializers/events/for_group_collection_serializer_spec.rb [hunk 13]
```diff
@@ -6,7 +6,7 @@
   subject { JSON.parse(serialization.to_json) }
 
   it 'contains relevant keys' do
-    expected_keys = %w[id title date]
+    expected_keys = %w[id title date description location date_and_time]
```
Comment (CamronLDNF) [inline spec/serializers/events/for_group_collection_serializer_spec.rb]: duplication of dates

File: spec/serializers/events/index_serializer_spec.rb [hunk 14]
```diff
@@ -6,7 +6,7 @@
   subject { JSON.parse(serialization.to_json) }
 
   it 'contains relevant keys' do
-    expected_keys = %w[id title date group]
+    expected_keys = %w[id title date description location date_and_time group]
```
Comment (CamronLDNF) [inline spec/serializers/events/index_serializer_spec.rb]: duplication of dates

File: db/schema.rb (hunk 6 above)
Comment (Jon1969Edwards) [reply db/schema.rb]: Done.

File: app/controllers/events_controller.rb (hunk 1 above)
Comment (Jon1969Edwards) [reply app/controllers/events_controller.rb]: Have updated to ':time'.

File: app/models/event.rb (hunk 2 above)
Comment (Jon1969Edwards) [reply app/models/event.rb]: done

File: app/serializers/events/for_group_collection_serializer.rb (hunk 3 above)
Comment (Jon1969Edwards) [reply app/serializers/events/for_group_collection_serializer.rb]: done

File: app/serializers/events/index_serializer.rb (hunk 4 above)
Comment (Jon1969Edwards) [reply app/serializers/events/index_serializer.rb]: done

File: db/migrate/20190204200935_add_columns_to_event.rb (hunk 5 above)
Comment (Jon1969Edwards) [reply db/migrate/20190204200935_add_columns_to_event.rb]: done

File: db/schema.rb (hunk 7 above)
Comment (Jon1969Edwards) [reply db/schema.rb]: done

File: spec/factories/events.rb (hunk 8 above)
Comment (Jon1969Edwards) [reply spec/factories/events.rb]: done

File: spec/models/event_spec.rb (hunk 9 above)
Comment (Jon1969Edwards) [reply spec/models/event_spec.rb]: done

File: spec/models/event_spec.rb (hunk 10 above)
Comment (Jon1969Edwards) [reply spec/models/event_spec.rb]: done

File: spec/serializers/events/for_group_collection_serializer_spec.rb (hunk 13 above)
Comment (Jon1969Edwards) [reply spec/serializers/events/for_group_collection_serializer_spec.rb]: Fixed.

File: spec/serializers/events/index_serializer_spec.rb (hunk 14 above)
Comment (Jon1969Edwards) [reply spec/serializers/events/index_serializer_spec.rb]: Fixed

File: spec/requests/organizer_can_create_event_spec.rb (hunk 12 above)
Comment (Jon1969Edwards) [reply spec/requests/organizer_can_create_event_spec.rb]: Organizer will already be associated through the 'group' so no need to associate him/her here.

File: spec/requests/organizer_can_create_event_spec.rb (hunk 11 above)
Comment (Jon1969Edwards) [reply spec/requests/organizer_can_create_event_spec.rb]: 1. fixed
2. Organizer will already be associated through the 'group' so no need to associate him/her here.
3. done

File: spec/factories/events.rb [hunk 15]
```diff
@@ -2,6 +2,9 @@
   factory :event do
     title { "EventName" }
     date { "2019-12-12" }
+    description { "Description for event" }
+    location { "Location for event" }
+    time { "12:12:12" }
```
Comment (CamronLDNF) [inline spec/factories/events.rb:7]: Is this a correct time format?

File: spec/requests/organizer_can_create_event_spec.rb (hunk 12 above)
Comment (CamronLDNF) [reply spec/requests/organizer_can_create_event_spec.rb]: Noted. Thanks for clarifying.

File: spec/requests/organizer_can_create_event_spec.rb [hunk 16]
```diff
@@ -2,16 +2,18 @@
...
     
     before do
-      post "/groups/#{group.id}/events", params: { event: {title: 'Craft Academy'} }, headers: headers
+      post "/groups/#{group.id}/events", params: { event: 
+                                                  {title: 'Craft Academy',
+                                                  description: 'Graduation Party',
+                                                  location: 'Stockholm',
+                                                  date: '2019-12-12',
+                                                  time: '2000-01-01T12:12:12.000Z'
```
Comment (CamronLDNF) [inline spec/requests/organizer_can_create_event_spec.rb]: Time format is incorrect. Look up an example.

File: spec/requests/user_can_get_event_with_all_attributes_spec.rb [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+  let(:user_credentials) { user.create_new_auth_token }
+  let(:headers) { { HTTP_ACCEPT: 'application/json' }.merge!(user_credentials) }
+  
+  before do
+    post "/groups/#{group.id}/events", params: { event: 
+                                                {title: 'Craft Academy',
+                                                description: 'Graduation Party',
+                                                location: 'Stockholm',
+                                                date: '2019-12-12',
+                                                time: '2000-01-01T12:12:12.000Z'
```
Comment (CamronLDNF) [inline spec/requests/user_can_get_event_with_all_attributes_spec.rb]: Time format is off

File: README.md [hunk 18]
```diff
@@ -19,11 +18,10 @@ The live site can be viewed [here]().
 * Organizer can create an event
 * User can create a group
 * Organizer can set Category for group
-
+* Organizer can set additional event info
```
Comment (tochman) [inline README.md]: What does that mean?

File: db/seeds.rb [hunk 19]
```diff
@@ -1,7 +1 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
-#
-# Examples:
-#
-#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
-#   Character.create(name: 'Luke', movie: movies.first)
+Event.create(title: 'STHLM TECH Meetup', description: 'tech meeting', location: 'Stockholm', date: '2019-12-12', time: '19:00', group_id: 1 )
```
Comment (tochman) [inline db/seeds.rb]: Is there a group with ID 1?

File: spec/requests/organizer_can_create_event_spec.rb (hunk 16 above)
Comment (tochman) [inline spec/requests/organizer_can_create_event_spec.rb]: ?? Why set time in year 2000?

File: spec/requests/user_can_get_event_with_all_attributes_spec.rb [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+                                                } }, headers: headers
+  
+    get '/events'
```
Comment (tochman) [inline spec/requests/user_can_get_event_with_all_attributes_spec.rb]: I don’t get this. Why are you making 2 calls. What are you testing here. The POST or the GET request?

**Review by tochman — CHANGES_REQUESTED** [review]

File: db/seeds.rb (hunk 19 above)
Comment (Jon1969Edwards) [reply db/seeds.rb]: This has been removed.

File: spec/requests/organizer_can_create_event_spec.rb (hunk 16 above)
Comment (Jon1969Edwards) [reply spec/requests/organizer_can_create_event_spec.rb]: We came across this when working with Oliver.  This is the output it gives us when we put in the correct format.  The test will only pass when we put it in THIS format.  Couldn't figure this one out.
screen shot 2019-02-05 at 8 32 27 pm

File: spec/factories/events.rb (hunk 15 above)
Comment (Jon1969Edwards) [reply spec/factories/events.rb:7]: Yes.

File: spec/requests/organizer_can_create_event_spec.rb (hunk 16 above)
Comment (Jon1969Edwards) [reply spec/requests/organizer_can_create_event_spec.rb]: I've addressed this in an above comment to Thomas.

File: spec/requests/user_can_get_event_with_all_attributes_spec.rb (hunk 17 above)
Comment (Jon1969Edwards) [reply spec/requests/user_can_get_event_with_all_attributes_spec.rb]: I've addressed this in an above comment to Thomas.

File: spec/requests/user_can_get_event_with_all_attributes_spec.rb (hunk 20 above)
Comment (shahin1290) [reply spec/requests/user_can_get_event_with_all_attributes_spec.rb]: Basically we are checking GET request here to make sure that user can get all the attributes back.

File: README.md (hunk 18 above)
Comment (Jon1969Edwards) [reply README.md]: Right.  Wasn't sure with this one but I guess it's not really a feature.  Has been deleted.

**Review by tochman — APPROVED** [review]

**Pull request merged**